File: src/utils/updateCheck.js
// Is there a newer Atlas than the one running?
//
// **The first thing in Atlas that talks to a server, and the README has to
// change with it.** Until now the only outbound request was a barcode lookup the
// user initiates. This one fires on its own, so it is worth being exact about
// what leaves the phone: a GET to GitHub's public releases API for a public
// repository, with no query string, no body, no identifier, and no account. It
// is the same request anyone loading the releases page in a browser makes.
//
// **Why it exists.** A user on 1.0.5 could not pair their strap because the app
// matched one exact Bluetooth name. That was fixed the same day, and the only way
// they would ever learn is if the author messaged them personally. Every future
// fix has the same problem, so the cost compounds.

import { load, persist } from "@/utils/storage";
import { publishAvailableVersion } from "@/utils/nativeSummary";

const LATEST_URL = "https://api.github.com/repos/atlas-healthapp/atlas/releases/latest";
const RELEASES_PAGE = "https://github.com/atlas-healthapp/atlas/releases/latest";

/** What was found last time, so a launch with no network still knows. */
const CACHE_KEY = "atlas_update_check";

/**
 * How long a result is trusted.
 *
 * GitHub allows 60 unauthenticated requests an hour per IP, which is not a
 * constraint for one call a day, but a release is not something that happens
 * hourly either. Six hours means a fix published in the morning is offered by the
 * evening without the app ever behaving like a poller.
 */
const MAX_AGE_MS = 6 * 60 * 60 * 1000;

/** Give up rather than hang a settings screen on a dead network. */
const TIMEOUT_MS = 8000;

/**
 * `1.0.6` -> [1, 0, 6]. Tolerates a `v` prefix, which is how the tags are named,
 * and drops any build suffix, which is how debug builds identify themselves
 * (`1.0.5+a1b2c3d-dev`). A dev build therefore compares as its base version, so
 * it is offered the same update a release would be.
 */
export function parseVersion(value) {
  if (typeof value !== "string") return null;
  const cleaned = value.trim().replace(/^v/i, "").split(/[+\-\s]/)[0];
  const parts = cleaned.split(".");
  if (!parts.length || parts.length > 4) return null;
  // **Digits only, and every part non-empty.** `Number("")` is 0, not NaN, so a
  // bare "" and a leading-dash "-1.0.0" (whose split leaves an empty first part)
  // both parsed as version [0] and would have been treated as older than
  // everything - meaning any garbage tag on the releases page would nag every
  // user forever. Caught by the test, not by reading it back.
  if (parts.some((p) => !/^\d+$/.test(p))) return null;
  return parts.map((p) => Number(p));
}

/**
 * Is `candidate` newer than `current`? Compares part by part rather than as a
 * number, or 1.0.10 would sort below 1.0.9.
 *
 * Unparseable input is never "newer". A malformed tag on the releases page must
 * not nag every user forever, and an unrecognised local version must not make
 * the app think it is out of date with itself.
 */
export function isNewer(candidate, current) {
  const a = parseVersion(candidate);
  const b = parseVersion(current);
  if (!a || !b) return false;
  const len = Math.max(a.length, b.length);
  for (let i = 0; i < len; i++) {
    const x = a[i] ?? 0;
    const y = b[i] ?? 0;
    if (x !== y) return x > y;
  }
  return false;
}

/** The cached answer, whatever its age. Used to paint before a fetch returns. */
export function cachedUpdate() {
  const cached = load(CACHE_KEY, null);
  if (!cached || typeof cached !== "object") return null;
  // A cached hit is only still a hit if this build is still older than it. After
  // updating, the stale entry would otherwise offer the version now running.
  if (!isNewer(cached.latest, __APP_VERSION__)) return null;
  return { latest: cached.latest, url: cached.url || RELEASES_PAGE, notes: cached.notes ?? "" };
}

/**
 * Hand the answer to native, so the background service can say so once.
 *
 * **Here rather than at the two call sites.** Home and Settings both ask, and a
 * mirror written by one of them is a mirror the other forgets: the same defect
 * `metricRegistry` and `resolveSessions` exist to prevent, one file over. This
 * runs wherever the question is answered, including from the cache, because a
 * launch with no network still knows what it knew yesterday and the service
 * should still be able to say it.
 *
 * Deliberately not awaited by callers and never allowed to throw. Whether there
 * is a newer Atlas is not worth failing a settings screen over, which is the
 * rule the whole module already runs on.
 */
function mirrorToNative(found) {
  if (!found?.latest) return;
  publishAvailableVersion(found.latest).catch(() => null);
}

/**
 * Ask GitHub, at most once every six hours.
 *
 * **Never throws and never rejects.** It is called from app startup and from a
 * settings panel, and no failure here is worth a visible error: not knowing
 * whether an update exists is the normal state, not a fault. Returns the same
 * shape as `cachedUpdate`, or null.
 */
export async function checkForUpdate({ force = false } = {}) {
  const cached = load(CACHE_KEY, null);
  const age = cached?.at ? Date.now() - cached.at : Infinity;
  if (!force && age < MAX_AGE_MS) {
    const found = cachedUpdate();
    mirrorToNative(found);
    return found;
  }

  let controller = null;
  let timer = null;
  try {
    controller = new AbortController();
    timer = setTimeout(() => controller.abort(), TIMEOUT_MS);
    const response = await fetch(LATEST_URL, {
      signal: controller.signal,
      headers: { Accept: "application/vnd.github+json" },
    });
    if (!response.ok) {
      const found = cachedUpdate();
      mirrorToNative(found);
      return found;
    }
    const body = await response.json();
    const latest = String(body?.tag_name ?? "").replace(/^v/i, "");
    if (!parseVersion(latest)) {
      const found = cachedUpdate();
      mirrorToNative(found);
      return found;
    }

    persist(CACHE_KEY, {
      at: Date.now(),
      latest,
      url: body?.html_url || RELEASES_PAGE,
      // Trimmed hard. The panel shows a line, not a changelog, and a release body
      // can run to several screens.
      notes: String(body?.body ?? "").slice(0, 400),
    });
    const found = cachedUpdate();
    mirrorToNative(found);
    return found;
  } catch {
    // Offline, aborted, rate limited, or GitHub having a bad day. The last known
    // answer is still the best one available.
    const found = cachedUpdate();
    mirrorToNative(found);
    return found;
  } finally {
    if (timer) clearTimeout(timer);
  }
}

export const RELEASES_URL = RELEASES_PAGE;

File: android/app/src/main/java/io/github/atlashealthapp/atlas/ble/UpdateCheck.java
package io.github.atlashealthapp.atlas.ble;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Is there a newer Atlas than the one running? The pure half.
 *
 * <p><b>The twin of {@code src/utils/updateCheck.js}, and change-both-or-neither.</b>
 * The app asks GitHub while it is open and shows an in-app flag; this service
 * asks while it is closed and posts a notification. Both have to agree about what
 * counts as newer, or the shade offers an update the app does not, or the other
 * way round. {@code updateCheckTwin.test.js} reads this file and checks the two
 * against the same cases.
 *
 * <p><b>Why the service asks at all, having said it should not.</b> The first
 * version read a version string the app mirrored into SharedPreferences, on the
 * argument that one implementation is better than two. That was right about the
 * implementation and wrong about the feature: the mirror is only written while
 * Atlas is open, so the notification could only ever tell you something you had
 * already been shown on Home. The person it exists for is the one who has not
 * opened Atlas in a month. So the service asks for itself, and the mirror stays
 * as the fallback for a tick whose own request failed.
 *
 * <p><b>No {@code org.json}.</b> Under plain JUnit that class is an android.jar
 * stub which throws on every call, so anything parsed with it cannot be tested -
 * the same reason {@link AlarmPlan} is flat text rather than JSON. The one field
 * needed here is a string in a flat object, and a regex reads it honestly.
 */
final class UpdateCheck {

    private UpdateCheck() {
    }

    /**
     * {@code "tag_name": "v1.0.9"} out of the releases API's reply.
     *
     * <p>Deliberately not a general JSON parser. It matches the first
     * {@code tag_name} whose value is a plain string, which is the only shape
     * GitHub returns for it, and returns null for anything else rather than
     * guessing. A body that has changed shape must read as "no answer", never as
     * a version.
     */
    static String tagFromJson(final String body) {
        if (body == null) {
            return null;
        }
        final Matcher m = Pattern
                .compile("\"tag_name\"\\s*:\\s*\"([^\"\\\\]*)\"")
                .matcher(body);
        if (!m.find()) {
            return null;
        }
        final String tag = m.group(1).trim();
        return tag.isEmpty() ? null : tag;
    }

    /**
     * The one spelling of a version that both sources have to agree on.
     *
     * <p><b>This is what stops the notification arriving twice.</b> A release is
     * named twice: the app mirrors it having already stripped the {@code v}, and
     * this service's own fetch reads the raw tag with it still on. Both are
     * announced through the same watermark, so if they spelled the same release
     * differently the watermark would miss and the shade would carry
     * "Atlas 1.0.9 is available" twice over.
     *
     * <p>Returns null for anything that is not a version, so a spelling that
     * cannot be canonicalised is never recorded as one.
     */
    static String normalise(final String value) {
        if (value == null) {
            return null;
        }
        final String trimmed = value.trim();
        final String withoutV = trimmed.replaceFirst("^[vV]", "");
        return parseVersion(withoutV) == null ? null : withoutV;
    }

    /**
     * {@code 1.0.9} to {@code [1, 0, 9]}, or null for anything that is not a
     * version.
     *
     * <p>Tolerates the {@code v} the tags carry, and drops a build suffix so a
     * debug build compares as its base version. Digits only and every part
     * non-empty: in the JS twin an empty string and a leading-dash both parsed as
     * version zero, which would have made any garbage tag newer than everything
     * and nagged every user forever.
     */
    static int[] parseVersion(final String value) {
        if (value == null) {
            return null;
        }
        String cleaned = value.trim();
        if (cleaned.startsWith("v") || cleaned.startsWith("V")) {
            cleaned = cleaned.substring(1);
        }
        // Everything from the first +, - or whitespace is a build suffix.
        final int cut = indexOfAny(cleaned, "+- \t");
        if (cut >= 0) {
            cleaned = cleaned.substring(0, cut);
        }
        if (cleaned.isEmpty()) {
            return null;
        }
        final String[] parts = cleaned.split("\\.", -1);
        if (parts.length == 0 || parts.length > 4) {
            return null;
        }
        final int[] out = new int[parts.length];
        for (int i = 0; i < parts.length; i++) {
            if (parts[i].isEmpty()) {
                return null;
            }
            for (int j = 0; j < parts[i].length(); j++) {
                if (!Character.isDigit(parts[i].charAt(j))) {
                    return null;
                }
            }
            try {
                out[i] = Integer.parseInt(parts[i]);
            } catch (final NumberFormatException e) {
                return null;
            }
        }
        return out;
    }

    private static int indexOfAny(final String s, final String chars) {
        for (int i = 0; i < s.length(); i++) {
            if (chars.indexOf(s.charAt(i)) >= 0) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Is {@code candidate} newer than {@code current}?
     *
     * <p>Compared part by part rather than as a number, or 1.0.10 sorts below
     * 1.0.9. Unparseable input on either side is never newer: a malformed tag
     * must not nag everybody forever, and a version this build cannot read about
     * itself must not make it think it is out of date with itself.
     */
    static boolean isNewer(final String candidate, final String current) {
        final int[] a = parseVersion(candidate);
        final int[] b = parseVersion(current);
        if (a == null || b == null) {
            return false;
        }
        final int len = Math.max(a.length, b.length);
        for (int i = 0; i < len; i++) {
            final int x = i < a.length ? a[i] : 0;
            final int y = i < b.length ? b[i] : 0;
            if (x != y) {
                return x > y;
            }
        }
        return false;
    }
}
